src/overlay/TxDemandQueue.cpp

```diff
@@ -0,0 +1,66 @@
+// Copyright 2022 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "overlay/TxDemandQueue.h"
+
+#include "ledger/LedgerManager.h"
+#include "main/Application.h"
+
+namespace stellar
+{
+
+using namespace std;
+
+TxDemandQueue::TxDemandQueue(Application& app) : mApp(app)
+{
+}
+
+size_t
+TxDemandQueue::size() const
+{
+    return mIncomingTxHashes.size() + mTxHashesToRetry.size();
+}
+
+void
+TxDemandQueue::appendHashesToRetryAndMaybeTrim(std::list<Hash>& list)
+{
+    mTxHashesToRetry.splice(mTxHashesToRetry.end(), list);
+    while (mTxHashesToRetry.size() >
+           mApp.getLedgerManager().getLastMaxTxSetSizeOps())
+    {
+        mTxHashesToRetry.pop_front();
+    }
+}
+
+void
+TxDemandQueue::queueAndMaybeTrim(Hash const& hash)
```

Comment:
Updated

Reply:
can someone please confirm that the selected policy here is to drop retries before any new hashes? This basically means that we'll bias towards newer transactions during spikes instead of hanging on to older ones. I think this is what we agreed on today, but I want to be 100% sure (this policy sounds good to me, as it's consistent with the general FIFO approach throughout the design).